blaze-client/src/main/scala/org/http4s/blaze/client/PoolManager.scala

```diff
@@ -201,78 +205,115 @@ private final class PoolManager[F[_], A <: Connection[F]](
     * @param key The Request Key For The Connection
     * @return An effect of NextConnection
     */
-  def borrow(key: RequestKey): F[NextConnection] =
-    F.async { callback =>
-      semaphore.permit.use { _ =>
-        if (!isClosed) {
-          def go(): F[Unit] =
-            getConnectionFromQueue(key).flatMap {
-              case Some(pooled) if pooled.conn.isClosed =>
-                F.delay(logger.debug(s"Evicting closed connection for $key: $stats")) *>
-                  decrConnection(key) *>
-                  go()
-
-              case Some(pooled) if pooled.borrowDeadline.exists(_.isOverdue()) =>
-                F.delay(
-                  logger.debug(s"Shutting down and evicting expired connection for $key: $stats")
-                ) *>
-                  decrConnection(key) *>
-                  F.delay(pooled.conn.shutdown()) *>
-                  go()
-
-              case Some(pooled) =>
-                F.delay(logger.debug(s"Recycling connection for $key: $stats")) *>
-                  F.delay(callback(Right(NextConnection(pooled.conn, fresh = false))))
-
-              case None if numConnectionsCheckHolds(key) =>
-                F.delay(
-                  logger.debug(s"Active connection not found for $key. Creating new one. $stats")
-                ) *>
-                  createConnection(key, callback)
-
-              case None if maxConnectionsPerRequestKey(key) <= 0 =>
-                F.delay(callback(Left(NoConnectionAllowedException(key))))
-
-              case None if curTotal == maxTotal =>
-                val keys = idleQueues.keys
-                if (keys.nonEmpty)
-                  F.delay(
-                    logger.debug(
-                      s"No connections available for the desired key, $key. Evicting random and creating a new connection: $stats"
-                    )
-                  ) *>
-                    F.delay(keys.iterator.drop(Random.nextInt(keys.size)).next()).flatMap {
-                      randKey =>
-                        getConnectionFromQueue(randKey).map(
-                          _.fold(
-                            logger.warn(s"No connection to evict from the idleQueue for $randKey")
-                          )(_.conn.shutdown())
+  def borrow(key: RequestKey): F[NextConnection] = {
+    val connectionBorrowF =
+      F.defer(
+        Ref[F]
+          .of(false)
+          .flatMap(ref =>
```

Comment:
`ref` could get a more descriptive name here.  I had to keep reading a while to understand what this was.